File: go/experiment.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// Package represents a Go module to test.
type Package struct {
	ImportPath   string
	VersionSpec  string
	ImportExample string // Actual import to use in test code
}

// List of popular Go packages to test.
// Note: For v2+ modules, the module path includes the version suffix.
var packages = []Package{
	{"github.com/gorilla/mux", "v1", "github.com/gorilla/mux"},
	{"github.com/gin-gonic/gin", "v1", "github.com/gin-gonic/gin"},
	{"github.com/stretchr/testify", "v1", "github.com/stretchr/testify/assert"},
	{"github.com/spf13/cobra", "v1", "github.com/spf13/cobra"},
	{"github.com/urfave/cli/v2", "v2", "github.com/urfave/cli/v2"},
	{"github.com/sirupsen/logrus", "v1", "github.com/sirupsen/logrus"},
	{"go.uber.org/zap", "v1", "go.uber.org/zap"},
	{"github.com/go-resty/resty/v2", "v2", "github.com/go-resty/resty/v2"},
	{"gopkg.in/yaml.v3", "v3", "gopkg.in/yaml.v3"},
	{"github.com/lib/pq", "v1", "github.com/lib/pq"},
	{"github.com/go-sql-driver/mysql", "v1", "github.com/go-sql-driver/mysql"},
	{"github.com/google/uuid", "v1", "github.com/google/uuid"},
	{"github.com/pkg/errors", "v0.9", "github.com/pkg/errors"},
	{"github.com/hashicorp/go-multierror", "v1", "github.com/hashicorp/go-multierror"},
	{"github.com/prometheus/client_golang", "v1", "github.com/prometheus/client_golang/prometheus"},
	{"github.com/gorilla/websocket", "v1", "github.com/gorilla/websocket"},
	{"github.com/julienschmidt/httprouter", "v1", "github.com/julienschmidt/httprouter"},
	{"github.com/dgrijalva/jwt-go", "v3", "github.com/dgrijalva/jwt-go"},
	{"golang.org/x/sync", "v0.10", "golang.org/x/sync/errgroup"},
	{"golang.org/x/crypto", "v0.31", "golang.org/x/crypto/bcrypt"},
	{"github.com/go-chi/chi/v5", "v5", "github.com/go-chi/chi/v5"},
	{"github.com/labstack/echo/v4", "v4", "github.com/labstack/echo/v4"},
	{"github.com/gomodule/redigo", "v1", "github.com/gomodule/redigo/redis"},
	{"github.com/elastic/go-elasticsearch/v7", "v7", "github.com/elastic/go-elasticsearch/v7"},
	{"github.com/aws/aws-sdk-go", "v1", "github.com/aws/aws-sdk-go/aws"},
	{"google.golang.org/grpc", "v1", "google.golang.org/grpc"},
}

// Go versions to test.
// The golang.org/dl naming changed around 1.21 to include patch version.
var goVersions = []string{
	"1.13", "1.14", "1.15", "1.16", "1.17", "1.18",
	"1.19", "1.20", "1.21.0", "1.22.0", "1.23.0", "1.24.0",
}

// Get the Go binary path for a version.
func goBinPath(version string) string {
	gopath := os.Getenv("GOPATH")
	if gopath == "" {
		home, _ := os.UserHomeDir()
		gopath = filepath.Join(home, "go")
	}
	return filepath.Join(gopath, "bin", fmt.Sprintf("go%s", version))
}

// Check if a Go version wrapper is installed.
func goVersionInstalled(version string) bool {
	binPath := goBinPath(version)
	_, err := os.Stat(binPath)
	return err == nil
}

// Install a Go version wrapper via golang.org/dl.
func installGoWrapper(version string) error {
	dlPath := fmt.Sprintf("golang.org/dl/go%s@latest", version)
	cmd := exec.Command("go", "install", dlPath)
	if output, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("install wrapper: %v: %s", err, output)
	}
	return nil
}

// Ensure a Go version is installed and downloaded.
func ensureGoVersion(version string) error {
	if !goVersionInstalled(version) {
		fmt.Printf("    Installing Go %s wrapper...\n", version)
		if err := installGoWrapper(version); err != nil {
			return err
		}
	}

	// Always try download - it's fast if already present.
	binPath := goBinPath(version)
	cmd := exec.Command(binPath, "download")
	cmd.Stdout = nil
	cmd.Stderr = nil
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("download SDK: %v", err)
	}

	return nil
}

// Update go.mod to use a specific Go version directive.
func updateGoModVersion(tmpDir, version string) error {
	goModPath := filepath.Join(tmpDir, "go.mod")
	content, err := os.ReadFile(goModPath)
	if err != nil {
		return err
	}

	lines := strings.Split(string(content), "\n")
	for i, line := range lines {
		if strings.HasPrefix(strings.TrimSpace(line), "go ") {
			lines[i] = fmt.Sprintf("go %s", version)
			break
		}
	}

	return os.WriteFile(goModPath, []byte(strings.Join(lines, "\n")), 0644)
}

// Test if a project compiles with a specific Go version.
func testGoVersion(tmpDir, version string) (bool, error) {
	// Ensure toolchain is installed and downloaded.
	if err := ensureGoVersion(version); err != nil {
		return false, err
	}

	binPath := goBinPath(version)

	// Clean go.sum to allow fresh resolution.
	os.Remove(filepath.Join(tmpDir, "go.sum"))

	// Update go.mod to match the version being tested.
	if err := updateGoModVersion(tmpDir, version); err != nil {
		return false, err
	}

	// Try mod tidy with this version.
	tidyCmd := exec.Command(binPath, "mod", "tidy")
	tidyCmd.Dir = tmpDir
	if _, err := tidyCmd.CombinedOutput(); err != nil {
		return false, nil
	}

	// Try build.
	buildCmd := exec.Command(binPath, "build", ".")
	buildCmd.Dir = tmpDir
	if err := buildCmd.Run(); err != nil {
		return false, nil
	}

	return true, nil
}

// ExperimentResult stores the result for a single package test.
type ExperimentResult struct {
	PackageName       string  `json:"package_name"`
	DependencySpec    string  `json:"dependency_spec"`
	ResolvedVersion   *string `json:"resolved_version"`
	OldestCompatible  *string `json:"oldest_compatible"`
	LatestCompatible  *string `json:"latest_compatible"`
	Error             *string `json:"error"`
}

func main() {
	args := os.Args[1:]

	if len(args) > 0 {
		// Single package mode
		packageName := args[0]
		fmt.Printf("Testing single package: %s\n", packageName)
		runSinglePackageExperiment(packageName)
	} else {
		// Full experiment mode
		fmt.Println("Starting Go dependency toolchain compatibility experiment")
		fmt.Printf("Testing %d packages\n", len(packages))
		runFullExperiment()
	}
}

func runFullExperiment() {
	results := []ExperimentResult{}

	// Test control case
	fmt.Println("\n=== Testing control case (no dependencies) ===")
	controlResult, err := testControlCase()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Control case failed: %v\n", err)
	} else {
		fmt.Printf("Control: oldest=%v, latest=%v\n",
			ptrString(controlResult.OldestCompatible),
			ptrString(controlResult.LatestCompatible))
		results = append(results, controlResult)
	}

	// Test each package
	for _, pkg := range packages {
		fmt.Printf("\n=== Testing %s ===\n", pkg.ImportPath)
		result, err := testPackage(pkg)
		if err != nil {
			fmt.Fprintf(os.Stderr, "%s failed: %v\n", pkg.ImportPath, err)
			errStr := err.Error()
			results = append(results, ExperimentResult{
				PackageName:    pkg.ImportPath,
				DependencySpec: pkg.VersionSpec,
				Error:          &errStr,
			})
		} else {
			fmt.Printf("%s: oldest=%v, latest=%v\n",
				pkg.ImportPath,
				ptrString(result.OldestCompatible),
				ptrString(result.LatestCompatible))
			results = append(results, result)
		}
	}

	// Write results
	jsonData, err := json.MarshalIndent(results, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to marshal results: %v\n", err)
		os.Exit(1)
	}

	err = os.WriteFile("results.json", jsonData, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write results.json: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("\n=== Results written to results.json ===")
}

func runSinglePackageExperiment(packageName string) {
	// Find package in list
	var pkg *Package
	for i := range packages {
		if packages[i].ImportPath == packageName {
			pkg = &packages[i]
			break
		}
	}

	if pkg == nil {
		fmt.Printf("Warning: '%s' not found in predefined list\n", packageName)
		fmt.Println("Testing anyway with version spec 'v1'...")
		pkg = &Package{
			ImportPath:   packageName,
			VersionSpec:  "v1",
			ImportExample: packageName,
		}
	}

	fmt.Printf("\n=== Testing %s (version spec: %s) ===\n", pkg.ImportPath, pkg.VersionSpec)

	result, err := testPackage(*pkg)
	if err != nil {
		fmt.Fprintf(os.Stderr, "\nFailed to test %s: %v\n", packageName, err)
		os.Exit(1)
	}

	fmt.Printf("\nResults for %s:\n", packageName)
	fmt.Printf("  Dependency spec: %s\n", result.DependencySpec)
	fmt.Printf("  Resolved version: %s\n", ptrString(result.ResolvedVersion))
	fmt.Printf("  Oldest compatible: %s\n", ptrString(result.OldestCompatible))
	fmt.Printf("  Latest compatible: %s\n", ptrString(result.LatestCompatible))

	// Write result to file
	jsonData, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to marshal result: %v\n", err)
		os.Exit(1)
	}

	filename := fmt.Sprintf("result-%s.json", filepath.Base(packageName))
	err = os.WriteFile(filename, jsonData, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write %s: %v\n", filename, err)
		os.Exit(1)
	}

	fmt.Printf("\n=== Result written to %s ===\n", filename)
}

func testControlCase() (ExperimentResult, error) {
	tmpDir, err := os.MkdirTemp("", "go-exp-control-")
	if err != nil {
		return ExperimentResult{}, err
	}
	defer os.RemoveAll(tmpDir)

	// Create go.mod
	goMod := `module control

go 1.18
`
	err = os.WriteFile(filepath.Join(tmpDir, "go.mod"), []byte(goMod), 0644)
	if err != nil {
		return ExperimentResult{}, err
	}

	// Create main.go
	mainGo := `package main

func main() {
	// Control case with no dependencies
}
`
	err = os.WriteFile(filepath.Join(tmpDir, "main.go"), []byte(mainGo), 0644)
	if err != nil {
		return ExperimentResult{}, err
	}

	oldest := findOldestCompatible(tmpDir, Package{})
	latest := stringPtr(goVersions[len(goVersions)-1])

	return ExperimentResult{
		PackageName:      "CONTROL",
		DependencySpec:   "none",
		OldestCompatible: oldest,
		LatestCompatible: latest,
	}, nil
}

func testPackage(pkg Package) (ExperimentResult, error) {
	tmpDir, err := os.MkdirTemp("", "go-exp-")
	if err != nil {
		return ExperimentResult{}, err
	}
	defer os.RemoveAll(tmpDir)

	// Create minimal go.mod without dependencies
	goMod := `module test

go 1.18
`
	err = os.WriteFile(filepath.Join(tmpDir, "go.mod"), []byte(goMod), 0644)
	if err != nil {
		return ExperimentResult{}, err
	}

	// Create main.go that imports the package
	mainGo := fmt.Sprintf(`package main

import (
	_ "%s"
)

func main() {
	// Import package to verify it compiles
}
`, pkg.ImportExample)

	err = os.WriteFile(filepath.Join(tmpDir, "main.go"), []byte(mainGo), 0644)
	if err != nil {
		return ExperimentResult{}, err
	}

	// Use go get to add the dependency with proper version resolution
	// This will automatically pick a version compatible with our Go version
	getCmd := exec.Command("go", "get", pkg.ImportExample+"@latest")
	getCmd.Dir = tmpDir
	getOutput, err := getCmd.CombinedOutput()
	if err != nil {
		return ExperimentResult{}, fmt.Errorf("go get failed: %v\n%s", err, getOutput)
	}

	// Get resolved version
	resolvedVersion, err := getResolvedVersion(tmpDir, pkg.ImportPath)
	if err != nil {
		return ExperimentResult{}, err
	}

	oldest := findOldestCompatible(tmpDir, pkg)
	latest := stringPtr(goVersions[len(goVersions)-1])

	return ExperimentResult{
		PackageName:      pkg.ImportPath,
		DependencySpec:   pkg.VersionSpec,
		ResolvedVersion:  &resolvedVersion,
		OldestCompatible: oldest,
		LatestCompatible: latest,
	}, nil
}

func getResolvedVersion(tmpDir, importPath string) (string, error) {
	// Run go mod download to resolve dependencies
	cmd := exec.Command("go", "mod", "download")
	cmd.Dir = tmpDir
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("go mod download failed: %v\n%s", err, output)
	}

	// Run go list to get the resolved version
	cmd = exec.Command("go", "list", "-m", importPath)
	cmd.Dir = tmpDir
	output, err = cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("go list failed: %v\n%s", err, output)
	}

	// Parse output: "module version"
	parts := strings.Fields(string(output))
	if len(parts) >= 2 {
		return parts[1], nil
	}

	return "", fmt.Errorf("could not parse version from go list output")
}

// Find the oldest compatible Go version using binary search.
func findOldestCompatible(tmpDir string, pkg Package) *string {
	left := 0
	right := len(goVersions)
	var oldest *string

	for left < right {
		mid := left + (right-left)/2
		version := goVersions[mid]

		fmt.Printf("  Testing Go %s\n", version)

		works, err := testGoVersion(tmpDir, version)
		if err != nil {
			fmt.Printf("    Error: %v\n", err)
			left = mid + 1
			continue
		}

		if works {
			fmt.Printf("    OK\n")
			oldest = stringPtr(version)
			right = mid
		} else {
			fmt.Printf("    Failed\n")
			left = mid + 1
		}
	}

	return oldest
}

// Find the latest compatible Go version.
func findLatestCompatible(tmpDir string) *string {
	// Start from newest and work backward.
	for i := len(goVersions) - 1; i >= 0; i-- {
		version := goVersions[i]
		works, err := testGoVersion(tmpDir, version)
		if err != nil {
			continue
		}
		if works {
			return stringPtr(version)
		}
	}
	return nil
}

func stringPtr(s string) *string {
	return &s
}

func ptrString(s *string) string {
	if s == nil {
		return "N/A"
	}
	return *s
}

File: node/experiment.js
#!/usr/bin/env node

/**
 * Dependency toolchain compatibility experiment for Node.js.
 *
 * This experiment tests how adding a single dependency affects the range
 * of Node.js versions that can successfully use that dependency.
 *
 * Uses nvm for Node.js version management and npm for package installation.
 * Two-phase testing: --engine-strict for declared compatibility, then runtime tests.
 */

const fs = require('fs');
const path = require('path');
const { execSync, spawnSync } = require('child_process');
const os = require('os');

// Package list with known engines declarations.
const PACKAGES = [
  // Web frameworks.
  { name: 'express', hasEngines: true },
  { name: 'koa', hasEngines: true },

  // Utilities.
  { name: 'lodash', hasEngines: false },
  { name: 'date-fns', hasEngines: false },
  { name: 'uuid', hasEngines: false },

  // HTTP clients.
  { name: 'axios', hasEngines: false },
  { name: 'node-fetch', hasEngines: true },

  // Testing.
  { name: 'jest', hasEngines: true },
  { name: 'mocha', hasEngines: true },
  { name: 'vitest', hasEngines: true },

  // Async.
  { name: 'async', hasEngines: false },

  // CLI.
  { name: 'commander', hasEngines: true },
  { name: 'yargs', hasEngines: true },
  { name: 'chalk', hasEngines: true },

  // File/Path.
  { name: 'fs-extra', hasEngines: true },
  { name: 'glob', hasEngines: true },

  // Validation.
  { name: 'joi', hasEngines: true },
  { name: 'zod', hasEngines: false },

  // Database.
  { name: 'pg', hasEngines: true },

  // Logging.
  { name: 'winston', hasEngines: true },
  { name: 'pino', hasEngines: false },
];

// Node versions to test (LTS versions + current).
const NODE_VERSIONS = [
  'v14.21.3',  // LTS Fermium (EOL April 2023)
  'v16.20.2',  // LTS Gallium (EOL Sept 2023)
  'v18.20.8',  // LTS Hydrogen (Active until Oct 2025)
  'v20.19.5',  // LTS Iron (Current Active LTS)
  'v22.20.0',  // LTS Jod (Newest LTS)
  'v24.0.2',   // Current
];

const results = [];

console.log('Starting Node.js dependency toolchain compatibility experiment');
console.log(`Testing ${PACKAGES.length} packages across ${NODE_VERSIONS.length} Node versions\n`);

// Test control case first.
console.log('=== Testing control case (no dependencies) ===');
const controlResult = testControlCase();
console.log(`Control: oldest=${controlResult.oldestCompatible}, latest=${controlResult.latestCompatible}\n`);
results.push(controlResult);

// Test each package.
for (const pkg of PACKAGES) {
  console.log(`=== Testing ${pkg.name} ===`);
  const result = testPackage(pkg);
  console.log(`${pkg.name}: oldest=${result.oldestCompatible || 'NONE'}, latest=${result.latestCompatible || 'N/A'}`);
  if (result.engineFailure) {
    console.log(`  Engine restriction: ${result.engineFailure}`);
  }
  console.log();
  results.push(result);
}

// Write results.
fs.writeFileSync('results.json', JSON.stringify(results, null, 2));
console.log('=== Results written to results.json ===');

/**
 * Test the control case (no dependencies).
 */
function testControlCase() {
  const result = {
    packageName: 'CONTROL',
    hasEngines: false,
    oldestCompatible: null,
    latestCompatible: null,
  };

  const oldest = findOldestCompatible(null);
  result.oldestCompatible = oldest;
  result.latestCompatible = NODE_VERSIONS[NODE_VERSIONS.length - 1];

  return result;
}

/**
 * Test a single package.
 */
function testPackage(pkg) {
  const result = {
    packageName: pkg.name,
    hasEngines: pkg.hasEngines,
    oldestCompatible: null,
    latestCompatible: null,
    engineFailure: null,
  };

  // Find oldest compatible version.
  const oldest = findOldestCompatible(pkg);
  if (oldest) {
    result.oldestCompatible = oldest;
    result.latestCompatible = NODE_VERSIONS[NODE_VERSIONS.length - 1];
  }

  return result;
}

/**
 * Find the oldest compatible Node version using binary search.
 */
function findOldestCompatible(pkg) {
  let left = 0;
  let right = NODE_VERSIONS.length;
  let oldest = null;

  while (left < right) {
    const mid = Math.floor(left + (right - left) / 2);
    const version = NODE_VERSIONS[mid];

    console.log(`  Testing Node ${version}`);

    const testResult = testNodeVersion(version, pkg);

    if (testResult.success) {
      oldest = version;
      right = mid;
    } else {
      if (testResult.engineFailure && !results.find(r => r.packageName === pkg?.name)?.engineFailure) {
        // Record first engine failure we encounter.
        const pkgResult = results.find(r => r.packageName === pkg?.name);
        if (pkgResult) {
          pkgResult.engineFailure = testResult.engineFailure;
        }
      }
      left = mid + 1;
    }
  }

  return oldest;
}

/**
 * Test if a package works with a specific Node version.
 * Two-phase approach: engine-strict check, then runtime test.
 */
function testNodeVersion(nodeVersion, pkg) {
  // Ensure Node version is installed.
  if (!ensureNodeInstalled(nodeVersion)) {
    return { success: false, reason: 'Node install failed' };
  }

  // Control case - just test Node itself.
  if (!pkg) {
    return testNodeOnly(nodeVersion);
  }

  // Phase 1: Try npm install with --engine-strict.
  const engineResult = testWithEngineStrict(nodeVersion, pkg);
  if (!engineResult.success) {
    return { success: false, engineFailure: engineResult.reason };
  }

  // Phase 2: Runtime test - actually use the package.
  const runtimeResult = testRuntime(nodeVersion, pkg);
  return runtimeResult;
}

/**
 * Test that Node itself works (control case).
 */
function testNodeOnly(nodeVersion) {
  try {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'node-exp-'));

    try {
      // Create simple test.
      const testCode = 'console.log("ok");';
      fs.writeFileSync(path.join(tmpDir, 'test.js'), testCode);

      // Run with this Node version.
      const result = spawnSync('bash', ['-c', `source ~/.config/nvm/nvm.sh && nvm use ${nodeVersion} && node test.js`], {
        cwd: tmpDir,
        timeout: 30000,
        encoding: 'utf-8',
      });

      return {
        success: result.status === 0 && result.stdout?.includes('ok'),
      };
    } finally {
      fs.rmSync(tmpDir, { recursive: true, force: true });
    }
  } catch (err) {
    return { success: false, reason: err.message };
  }
}

/**
 * Phase 1: Test with npm install --engine-strict.
 */
function testWithEngineStrict(nodeVersion, pkg) {
  try {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'node-exp-'));

    try {
      // Create package.json.
      fs.writeFileSync(path.join(tmpDir, 'package.json'), JSON.stringify({
        name: 'test',
        version: '1.0.0',
        dependencies: {
          [pkg.name]: 'latest',
        },
      }, null, 2));

      // Create .npmrc with engine-strict=true.
      fs.writeFileSync(path.join(tmpDir, '.npmrc'), 'engine-strict=true\n');

      // Try npm install.
      const result = spawnSync('bash', ['-c',
        `source ~/.config/nvm/nvm.sh && nvm use ${nodeVersion} && npm install 2>&1`
      ], {
        cwd: tmpDir,
        timeout: 120000,
        encoding: 'utf-8',
      });

      if (result.status !== 0) {
        // Check if it's an engine error.
        const output = result.stdout + result.stderr;
        if (output.includes('EBADENGINE') || output.includes('Unsupported engine')) {
          // Extract the engine requirement if possible.
          const match = output.match(/wanted: \{[^}]+node[^}]+\}/);
          return {
            success: false,
            reason: match ? match[0] : 'Engine version mismatch',
          };
        }
        return { success: false, reason: 'Install failed (non-engine)' };
      }

      return { success: true };
    } finally {
      fs.rmSync(tmpDir, { recursive: true, force: true });
    }
  } catch (err) {
    return { success: false, reason: err.message };
  }
}

/**
 * Phase 2: Runtime test - actually require and use the package.
 */
function testRuntime(nodeVersion, pkg) {
  try {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'node-exp-'));

    try {
      // Create package.json (without engine-strict this time).
      fs.writeFileSync(path.join(tmpDir, 'package.json'), JSON.stringify({
        name: 'test',
        version: '1.0.0',
        dependencies: {
          [pkg.name]: 'latest',
        },
      }, null, 2));

      // Install package.
      const installResult = spawnSync('bash', ['-c',
        `source ~/.config/nvm/nvm.sh && nvm use ${nodeVersion} && npm install --silent 2>&1`
      ], {
        cwd: tmpDir,
        timeout: 120000,
        encoding: 'utf-8',
      });

      if (installResult.status !== 0) {
        return { success: false, reason: 'Runtime install failed' };
      }

      // Create test code that actually uses the package.
      const testCode = generateTestCode(pkg.name);
      fs.writeFileSync(path.join(tmpDir, 'test.js'), testCode);

      // Run the test.
      const testResult = spawnSync('bash', ['-c',
        `source ~/.config/nvm/nvm.sh && nvm use ${nodeVersion} && node test.js`
      ], {
        cwd: tmpDir,
        timeout: 30000,
        encoding: 'utf-8',
      });

      return {
        success: testResult.status === 0 && testResult.stdout?.includes('TEST_OK'),
      };
    } finally {
      fs.rmSync(tmpDir, { recursive: true, force: true });
    }
  } catch (err) {
    return { success: false, reason: err.message };
  }
}

/**
 * Generate test code that actually uses a package.
 * This avoids the Java experiment's mistake of testing with an empty file.
 */
function generateTestCode(packageName) {
  const tests = {
    'express': `
      const express = require('express');
      const app = express();
      app.get('/', (req, res) => res.send('ok'));
      console.log('TEST_OK');
    `,
    'koa': `
      const Koa = require('koa');
      const app = new Koa();
      console.log('TEST_OK');
    `,
    'lodash': `
      const _ = require('lodash');
      const result = _.chunk(['a', 'b', 'c', 'd'], 2);
      if (result.length === 2) console.log('TEST_OK');
    `,
    'axios': `
      const axios = require('axios');
      console.log(typeof axios.get === 'function' ? 'TEST_OK' : 'FAIL');
    `,
    'uuid': `
      const { v4 } = require('uuid');
      const id = v4();
      console.log(id.length === 36 ? 'TEST_OK' : 'FAIL');
    `,
    'chalk': `
      const chalk = require('chalk');
      console.log(chalk.blue ? 'TEST_OK' : 'FAIL');
    `,
    'commander': `
      const { Command } = require('commander');
      const program = new Command();
      console.log('TEST_OK');
    `,
    'winston': `
      const winston = require('winston');
      const logger = winston.createLogger({transports: [new winston.transports.Console({silent: true})]});
      console.log('TEST_OK');
    `,
    'glob': `
      const { glob } = require('glob');
      console.log(typeof glob === 'function' ? 'TEST_OK' : 'FAIL');
    `,
    'fs-extra': `
      const fs = require('fs-extra');
      console.log(typeof fs.ensureDir === 'function' ? 'TEST_OK' : 'FAIL');
    `,
  };

  // Return specific test if available, otherwise generic require test.
  return tests[packageName] || `
    const pkg = require('${packageName}');
    console.log(pkg ? 'TEST_OK' : 'FAIL');
  `;
}

/**
 * Ensure a Node version is installed via nvm.
 */
function ensureNodeInstalled(version) {
  try {
    // Check if already installed.
    const listResult = spawnSync('bash', ['-c', `source ~/.config/nvm/nvm.sh && nvm list`], {
      encoding: 'utf-8',
      timeout: 10000,
    });

    if (listResult.stdout?.includes(version)) {
      return true;
    }

    // Install the version.
    console.log(`    Installing Node ${version}...`);
    const installResult = spawnSync('bash', ['-c',
      `source ~/.config/nvm/nvm.sh && nvm install ${version}`
    ], {
      timeout: 300000,
      encoding: 'utf-8',
    });

    return installResult.status === 0;
  } catch (err) {
    console.log(`    Failed to install ${version}: ${err.message}`);
    return false;
  }
}
